### PR TITLE
Make ledger-tool list-roots subcommand inclusive of start-root arg

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
The `list-roots` ledger-tool subcommand has an off-by-one error in regards to which roots it will output.

- The old find-roots logic called `.enumerate()` and took entries until `i > num_roots`
    - This is equivalent to `i <= num_roots` and would yield `num_roots + 1` elements if available
- The output logic also called `.enumerate()` and output entries while `i < num_roots`

There is obviously an incongruence here in that we're collecting `num_roots + 1` entries and outputting `num_roots` of them; since we print the roots in descending order, this means we are dropping the oldest root (ie the `start-slot` if `start-slot` is indeed a root) from output.

#### Summary of Changes
Use "rustier" iterator conventions which eliminate the aforementioned bug.

### CODE REVIEW

**Comment by steviez** [comment]
Running the following command
```
solana-ledger-tool list-roots --ledger ~/ledger/ --start-root 166881914 --num-roots 3
```
Noting that [166881914](https://explorer.solana.com/block/166881914) is a root, here is the existing behavior (v1.14.7)
```
166881917: 8o8EwaZfJrFNKxYKnQMry2VyDSaFMcEeCVdhfUuda2jR
166881916: GsmQ6N2kAnz5jcKd5y6rkeuqFyAgmM3jWDYaGVvrsPwB
166881915: Hfj8NxeWdsVQpQFgWfKzvtmaJgTxHMTvTJH5N62EWvf8
```
vs. the output from my branch
```
166881916: GsmQ6N2kAnz5jcKd5y6rkeuqFyAgmM3jWDYaGVvrsPwB
166881915: Hfj8NxeWdsVQpQFgWfKzvtmaJgTxHMTvTJH5N62EWvf8
166881914: 4G2StaGe2V6DX1F6dn9ctPV6XXdAsL8iGvTM4RHAbwNs
```

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -4104,16 +4086,24 @@ fn main() {
...
+                    .map(|slot| {
+                        let blockhash = blockstore
+                            .get_slot_entries(slot, 0)
+                            .unwrap()
+                            .last()
+                            .unwrap()
+                            .hash;
+                        (slot, blockhash)
+                    })
+                    .collect::<Vec<_>>()
```
Comment (steviez) [inline ledger-tool/src/main.rs:4100]: This `.collect()` is necessary; removing it results in following compile errors (with or without `.into_iter()`)
```
error[E0277]: the trait bound `impl Iterator<Item = u64>: DoubleEndedIterator` is not satisfied
    --> ledger-tool/src/main.rs:4089:17
     |
4089 | /                 iter.take(num_roots)
4090 | |                     .filter(|slot| *slot <= max_height as u64)
4091 | |                     .map(|slot| {
4092 | |                         let blockhash = blockstore
...    |
4098 | |                         (slot, blockhash)
4099 | |                     })
     | |______________________^ the trait `DoubleEndedIterator` is not implemented for `impl Iterator<Item = u64>`
4100 |                       .rev()
     |                        --- required by a bound introduced by this call
     |
     = help: the trait `DoubleEndedIterator` is implemented for `std::iter::Map<I, F>`
     = note: required for `std::iter::Take<impl Iterator<Item = u64>>` to implement `DoubleEndedIterator`
     = note: 2 redundant requirements hidden
     = note: required for `std::iter::Map<Filter<std::iter::Take<impl Iterator<Item = u64>>, [closure@ledger-tool/src/main.rs:4090:29: 4090:35]>, [closure@ledger-tool/src/main.rs:4091:26: 4091:32]>` to implement `DoubleEndedIterator`
note: required by a bound in `std::iter::Iterator::rev`

error[E0277]: the trait bound `impl Iterator<Item = u64>: ExactSizeIterator` is not satisfied
    --> ledger-tool/src/main.rs:4089:17
     |
4089 | /                 iter.take(num_roots)
4090 | |                     .filter(|slot| *slot <= max_height as u64)
4091 | |                     .map(|slot| {
4092 | |                         let blockhash = blockstore
...    |
4098 | |                         (slot, blockhash)
4099 | |                     })
     | |______________________^ the trait `ExactSizeIterator` is not implemented for `impl Iterator<Item = u64>`
4100 |                       .rev()
     |                        --- required by a bound introduced by this call
     |
     = help: the following other types implement trait `ExactSizeIterator`:
               &mut I
               Args
               ArgsOs
               ArrayChunksMut<'_, T, N>
               ArrayWindows<'_, T, N>
               Box<I, A>
               ByteRecordIter<'r>
               ChunksExact<'_, T>
             and 274 others
     = note: required for `std::iter::Take<impl Iterator<Item = u64>>` to implement `DoubleEndedIterator`
     = note: 2 redundant requirements hidden
     = note: required for `std::iter::Map<Filter<std::iter::Take<impl Iterator<Item = u64>>, [closure@ledger-tool/src/main.rs:4090:29: 4090:35]>, [closure@ledger-tool/src/main.rs:4091:26: 4091:32]>` to implement `DoubleEndedIterator`
note: required by a bound in `std::iter::Iterator::rev`
```

**Review by ilya-bobyr — DISMISSED** [review]
That is a pretty tricky bug!
New logic looks correct to me.

After reading the code, I figured, slot processing can be done in a single loop, more or less.  Common `Iterator` functions can be used to do the filtering specified by the arguments.    And then for argument parsing, one can use a `map()`...

So, I've rewritten the whole function %)

```rust
let blockstore = open_blockstore(
    &ledger_path,
    AccessType::Secondary,
    wal_recovery_mode,
    &shred_storage_type,
    force_update_to_open,
);
let max_height = arg_matches
    .value_of("max_height")
    .map(|v| u64::from_str(v).expect("Maximum height must be a number"))
    .unwrap_or(u64::MAX);
let start_root = arg_matches
    .value_of("start_root")
    .map(|v| Slot::from_str(v).expect("Starting root must be a number"))
    .unwrap_or(0);
let num_roots = arg_matches
    .value_of("num_roots")
    .map(|v| usize::from_str(v).expect("Number of roots must be a number"))
    .unwrap_or(usize::from_str(DEFAULT_ROOT_COUNT).unwrap());

let output_path = arg_matches.value_of("slot_list");
let mut output = output_path
    .map(|path| {
        Box::new(
            File::create(path)
                .expect(&format!("Failed to open \"{path}\" for output")),
        ) as Box<dyn Write>
    })
    .unwrap_or_else(|| Box::new(stdout()));

let rooted_slots = blockstore
    .rooted_slot_iterator(start_root)
    .expect("Failed to get rooted slot");

for slot in rooted_slots
    .take(num_roots)
    .take_while(|slot| *slot <= max_height)
    // We get entries starting with lower slot numbers, but we want to output
    // starting from the most recent ones.
    .collect::<Vec<_>>()
    .into_iter()
    .rev()
{
    let hash = blockstore
        .get_slot_entries(slot, 0)
        .unwrap()
        .last()
        .unwrap()
        .hash;

    writeln!(output, "{:?}: {:?}", slot, hash).expect(&format!(
        "Failed to write into \"{}\"",
        output_path.unwrap_or("-")
    ));
}
```

See for yourself if you can use any of this.
The change looks good either way.

By the way, there are some libraries for parsing CLI arguments that can do type conversions, default values and range checks.  [`clap`](https://docs.rs/clap/latest/clap/) seems like a pretty good choice.

File: ledger-tool/src/main.rs (hunk 1 above)
Comment (ilya-bobyr) [reply ledger-tool/src/main.rs:4100]: Checked that as well.
Iterator returned by the `open_blockstore` is just an [`Iterator`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html), not a [`DoubleEndedIterator`](https://doc.rust-lang.org/stable/std/iter/trait.DoubleEndedIterator.html) that [`rev()`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.rev) needs.
See the `rev()` method bounds:

```rust
    fn rev(self) -> Rev<Self>
    where
        Self: Sized + DoubleEndedIterator,
```

Seems reasonable, as `Iteator` only provides an ability to get items one by one and only once.
So one would need to either store them, as `collect::<Vec<_>>()` would do, or provide reverse iteration in some other way.

Cursory investigation shows, `open_blockstore` binds to the RocksDB API.
And there exists a way to iterate in reverse, via `reversed_rooted_slot_iterator()`, but then one needs to provide an end point.
And that seems to be harder to calculate, given the current tool arguments.
Guess, it is probably good enough as it is :)

**Comment by steviez** [comment]
> After reading the code, I figured, slot processing can be done in a single loop, more or less. Common Iterator functions can be used to do the filtering specified by the arguments

Sure, the code you suggested looks pretty similar to what I have, although `take_while()` (what you used) is more efficient than `filter()` (what I used) given that the slots will be increasing order. That is, once we find a slot that fails to satisfy `slot <= max_height as u64`, all subsequent slots will as well. I'll throw this in.

> And then for argument parsing, one can use a map()...

Yep, good point. I'm not going to make that change in this PR for the sake of doing one thing per PR/commit, but what you posted definitely looks cleaner. Feel free to do a quick PR if you'd like; if not, I can take a pass next time I have a little downtime.

> By the way, there are some libraries for parsing CLI arguments that can do type conversions, default values and range checks. [clap](https://docs.rs/clap/latest/clap/) seems like a pretty good choice.

Yeah, I think we use some functionality from `clap`, but I think some of the code in this tool might predate when we started using `clap`, or possibly before functionality was in `clap`? I don't have the full history, but there is almost assuredly code that can be improved as the language & supported syntax / ecosystem libraries have improved

**Comment by ilya-bobyr** [comment]
> > And then for argument parsing, one can use a map()...
> 
> Yep, good point. I'm not going to make that change in this PR for the sake of doing one thing per PR/commit, but what you posted definitely looks cleaner.

Absolutely, makes sense.

> Feel free to do a quick PR if you'd like; if not, I can take a pass next time I have a little downtime.
>
> > By the way, there are some libraries for parsing CLI arguments that can do type conversions, default values and range checks. [clap](https://docs.rs/clap/latest/clap/) seems like a pretty good choice.
> 
> Yeah, I think we use some functionality from `clap`, but I think some of the code in this tool might predate when we started using `clap`, or possibly before functionality was in `clap`? I don't have the full history, but there is almost assuredly code that can be improved as the language & supported syntax / ecosystem libraries have improved

Indeed, sees like `clap` is already used here.
I guess, it then makes sense to do some cleanup, as `clap` should be able to process arguments better than doing it “manually”.
I can also see that `clap` is used without the `derive` feature.
Allowing `clap` to write code based on a `struct` definition should remove some syntactic noise.
I can already see sub-commands defined – they should go into individual modules.
Each module defining necessary arguments as a struct with some annotations.
`clap` can then generate all the parsing code and combine those into a larger CLI.
I can do a pass, maybe a bit later.
But, I'm trying to fix some bugs assigned to be right now, and as I am still learning it takes time.

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -4104,16 +4086,22 @@ fn main() {
...
-                        }
+                    .for_each(|slot| {
+                        let blockhash = blockstore
+                            .get_slot_entries(slot, 0)
+                            .unwrap()
+                            .last()
+                            .unwrap()
+                            .hash;
+
+                        output_file
```
Comment (ilya-bobyr) [inline ledger-tool/src/main.rs:4102]: I've mostly seen this written using `writeln!`, rather than `.write_all(format!())`.

**Review by ilya-bobyr — APPROVED** [review]

**Comment by steviez** [comment]
> can do a pass, maybe a bit later.

Yep, definitely no rush on cleanup tasks 😄

File: ledger-tool/src/main.rs (hunk 2 above)
Comment (steviez) [reply ledger-tool/src/main.rs]: Fair - initially left it alone as this code didn't change but doubled back and cleaned it up

**Comment by steviez** [comment]
Merged without letting CI finish - this previously had all green, and passed sanity/checks. I'll fix it if my eagerness breaks shit

**Pull request merged**